### PR TITLE
Shell: Changes needed to enable shell reinitialization

### PR DESCRIPTION
**Pull request by nordic-krch**
### shell: Add callback to shell_uninit function
    
Shell uninitialization is not synchronous, it is deferred to shell
thread so resources used by the shell (e.g. transport resource like
uart) cannot be used until it is completed. Added callback which
notifies when all resources are released and shell is uninitialized.
Callback is called from shell thread just before it is aborted.

### shell: uart: Fix init function
    
Init function can be called multiple times (after each shell
reinitialization). It was missing reseting ring buffers and tx_busy
flag. When called once proper state of those variable where handled
by ram sections initialization.

### samples: shell_module: Add shell_uart_release command
    
shell_uart_release command shows how shell can be uninitialized and
release transport resources (uart) and reinitialized at runtime.

### CODE REVIEW

File: include/shell/shell.h [hunk 1]
```diff
@@ -485,6 +485,7 @@ enum shell_state {
 	SHELL_STATE_UNINITIALIZED,
 	SHELL_STATE_INITIALIZED,
 	SHELL_STATE_ACTIVE,
+	SHELL_STATE_HALTED,
```
Comment (jakub-uC) [inline include/shell/shell.h:488]: What is purpose of this new state?

Comment (yperess) [reply include/shell/shell.h:488]: Agreed, I think at this point a state diagram/transition list would be helpful. Can we only move to halted from active? Where can we go from the various panic modes? etc.

File: samples/subsys/shell/shell_module/src/uart_reinit.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	static struct k_work shell_init_work;
+
+	k_work_init(&shell_init_work, shell_init_from_work);
+	int err = k_work_submit(&shell_init_work);
+
+	(void)err;
+	__ASSERT_NO_MSG(err >= 0);
+}
+
+static void direct_uart_callback(const struct device *dev, void *user_data)
```
Comment (yperess) [inline samples/subsys/shell/shell_module/src/uart_reinit.c:34]: What if `CONFIG_UART_INTERRUPT_DRIVEN` isn't defined, this function seems to heavily rely on interrupts. Could we get away with calling poll_in and poll_out instead?

File: include/shell/shell.h (hunk 1 above)
Comment (nordic-krch) [reply include/shell/shell.h]: Removed.
> Where can we go from the various panic modes? 

Panic modes are entered when `LOG_PANIC` is called and shell as a logger backend is informed about panic. `LOG_PANIC` is called on unrecoverable system fault. There is no way back from panic.

File: samples/subsys/shell/shell_module/src/uart_reinit.c (hunk 2 above)
Comment (nordic-krch) [reply samples/subsys/shell/shell_module/src/uart_reinit.c:34]: I've updated sample to support case when shell uart is not using interrupts.

File: subsys/shell/shell.c [hunk 3]
```diff
@@ -1314,20 +1319,25 @@ int shell_init(const struct shell *shell, const void *transport_config,
...
 		/* signal kill message */
 		(void)k_poll_signal_raise(signal, 0);
 
-		return 0;
-	} else {
-		return instance_uninit(shell);
+		return;
+	}
+
+	int err = instance_uninit(shell);
```
Comment (jakub-uC) [inline subsys/shell/shell.c:1337]: Just a minor remark. Shall we handle this `err` ?

**Review by jakub-uC — APPROVED** [review]

**Review by Mierunski — APPROVED** [review]

**Pull request merged**